File: ModuleM/src/main/java/com/lib/monitor/methodcost/normal/MethodCostUtil.java
package com.lib.monitor.methodcost.normal;

import android.util.Log;

import java.util.concurrent.ConcurrentHashMap;

public class MethodCostUtil {
    private static ConcurrentHashMap<String, Long> methodCostMap = new ConcurrentHashMap<>();

    public static synchronized void recordObjectMethodCostStart(int thresholdTime, String methodName) {
        if (!Thread.currentThread().getName().equals("main")) {
            return;
        }
        methodCostMap.put(methodName, System.currentTimeMillis());
    }


    public static synchronized void recordObjectMethodCostEnd(int thresholdTime, String methodName) {
        if (!Thread.currentThread().getName().equals("main")) {
            return;
        }
        if (!methodCostMap.containsKey(methodName)) {
            return;
        }

        long starttime = methodCostMap.get(methodName);
        long cost = System.currentTimeMillis() - starttime;
        if (cost < thresholdTime) {
            methodCostMap.remove(methodName);
            return;
        }
        Log.v("AndroidTest", methodName + " cost = " + cost + "ms");
    }
}

File: ModuleM/src/main/java/com/didichuxing/doraemonkit/aop/MethodCostUtil.kt
package com.didichuxing.doraemonkit.aop

import android.util.Log
import java.lang.Exception
import java.util.concurrent.ConcurrentHashMap

object MethodCostUtil {
    const val TAG = "AndroidTest"

    private val METHOD_COSTS: ConcurrentHashMap<String, Long?> by lazy {
        ConcurrentHashMap<String, Long?>()
    }

    @Synchronized
    fun recordObjectMethodCostStart(thresholdTime: Int, methodName: String, classObj: Any?) {
        if (!Thread.currentThread().name.equals("main", true)) {
            return
        }
        try {
            METHOD_COSTS[methodName] = System.currentTimeMillis()
        } catch (e: Exception) {
            println("e = ${e.localizedMessage}")
        }
    }

    fun recordObjectMethodCostEnd(thresholdTime: Int, methodName: String, classObj: Any?) {
        if (!Thread.currentThread().name.equals("main", true)) {
            return
        }
        synchronized(MethodCostUtil::class.java) {
            try {
                if (METHOD_COSTS.containsKey(methodName)) {
                    val startTime = METHOD_COSTS[methodName]!!
                    val costTime = (System.currentTimeMillis() - startTime).toInt()
                    METHOD_COSTS.remove(methodName)
//                    if (costTime >= thresholdTime) {
                        val threadName = Thread.currentThread().name
                        Log.v(TAG, "================Dokit================")
                        Log.v(TAG, "\tmethodName===>$methodName threadName===>$threadName thresholdTime===>$thresholdTime costTime===>$costTime")
                        val stackTraceElements = Thread.currentThread().stackTrace
                        for (stackTraceElement in stackTraceElements) {
                            if (stackTraceElement.toString().contains("MethodCostUtil")) {
                                continue
                            }
                            if (stackTraceElement.toString().contains("dalvik.system.VMStack.getThreadStackTrace")) {
                                continue
                            }
                            if (stackTraceElement.toString().contains("java.lang.Thread.getStackThread")) {
                                continue
                            }
                            Log.v(TAG, "\tat $stackTraceElement")
//                        }
                    }
                }
            } catch (e: Exception) {
//                Log.v(TAG, "e = ${e.localizedMessage}")
            }
        }
    }
}